packages/amplify-cli-core/src/errors/amplify-exception.ts

```diff
@@ -1,27 +1,53 @@
 /**
  * Base class for all Amplify exceptions
  */
-export class AmplifyException extends Error {
+export abstract class AmplifyException extends Error {
   public readonly message: string;
   public readonly resolution?: string;
   public readonly details?: string;
   public readonly link?: string;
 
+  /**
+   * You should use AmplifyError or AmplifyFault to throw an exception.
+   *
+   * @param {AmplifyExceptionType} name - a user friendly name for the exception
+   * @param {AmplifyExceptionClassification} classification - Fault or Error
+   * @param {AmplifyExceptionOptions} options - error stack, resolution steps, details, or help links
+   * @param {Error | null} downstreamException If you are throwing this exception from within a catch block,
+   * you must provide the exception that was caught.
+   * @example
+   * try {
+   *  ...
+   * } catch (downstreamException){
+   *    throw new AmplifyError(...,...,downstreamException);
+   * }
+   */
   constructor(
     public readonly name: AmplifyExceptionType,
     public readonly classification: AmplifyExceptionClassification,
     private readonly options: AmplifyExceptionOptions,
+    public readonly downstreamException?: Error,
   ) {
-    super(options.message);
+    // If an AmplifyException was already thrown, we must allow it to reach the user.
+    // This ensures that resolution steps, and the original error are bubbled up.
+    super(downstreamException instanceof AmplifyException ? downstreamException.message : options.message);
 
     // https://github.com/Microsoft/TypeScript-wiki/blob/main/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
     Object.setPrototypeOf(this, AmplifyException.prototype);
 
-    this.stack = options.stack ?? this.stack;
-    this.message = options.message;
-    this.details = options.details;
-    this.resolution = 'resolution' in options ? options.resolution : undefined;
-    this.link = 'link' in options ? options.link : undefined;
+    if (downstreamException instanceof AmplifyException) {
+      this.stack = downstreamException.stack ? downstreamException.stack : this.stack;
+      this.message = downstreamException.message;
+      this.details = downstreamException.details;
+      this.resolution = downstreamException.resolution;
+      this.link = downstreamException.link;
+    } else {
+      this.stack = options.stack ? options.stack : this.stack;
```

Comment:
if the downstreamException is not `instanceof AmplifyException` can we get the stack and message to details from it, this will simplify the code where we have something like
```
throw new AmplifyError('DeploymentError', {
  message: 'Failed to deploy Auth and Pinpoint resources.',
  resolution: 'Deploy the Auth and Pinpoint resources manually.',
  details: err.message,
  stack: err.stack,
});
```
to
```
throw new AmplifyError('DeploymentError', {
  message: 'Failed to deploy Auth and Pinpoint resources.',
  resolution: 'Deploy the Auth and Pinpoint resources manually.',
}, err);
```

Reply:
good note for follow up!